### PR TITLE
[spec] Specify object lifecycle and related operations

### PR DESCRIPTION
**Pull request by thetorpedodog**
This formalizes the object lifecycle and the various creation, opening, and closing methods we are already using in the tiledbsoma Python implementation.

See #107.

### CODE REVIEW

**Review by bkmartinjr — DISMISSED** [review]
Caveat: I did a very fast review.  But based upon that, it looks good to me.

File: abstract_specification.md
Comment (mlin) [inline abstract_specification.md:418]:
```suggestion
All such objects will be automatically closed together by [the collection's close operation](#operation-close-collection-types).
```

**Review by mlin — APPROVED** [review]
LGTM, one minor language tightening suggested inline

File: abstract_specification.md
Comment (thetorpedodog) [reply abstract_specification.md]: This is good.

File: abstract_specification.md
Comment (johnkerl) [inline abstract_specification.md:513]: ditto re end w/ period or not

**Review by johnkerl — CHANGES_REQUESTED** [review]

File: abstract_specification.md
Comment (bkmartinjr) [reply abstract_specification.md:251]: For clarity, our internal specs tend to always use the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) standard (i.e., consistently use `NOT`), but given that this doc has not already adopted it, that is probably an aspirational change.

Comment (thetorpedodog) [reply abstract_specification.md:251]: an author [OUGHT TO](https://www.rfc-editor.org/rfc/rfc6919#section-4) use the caps consistently

Comment (johnkerl) [reply abstract_specification.md:251]: @bkmartinjr TIL about RFC2119! :) 

If we want to go that route, sure.

Comment (bkmartinjr) [reply abstract_specification.md:251]: there are very good reasons for it - which is why the IETF uses it (as does many other orgs like W3C).  It is a future decision which should not hold up this PR -- but coincidentally the topic was it was just raised inside CZI (regarding this very doc being unclear).  

[Here is](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md) an example of one of our specs that strongly benefits from it.

Comment (johnkerl) [reply abstract_specification.md:251]: I filed #122 -- this is awesome info -- thanks @bkmartinjr !! :)

**Review by thetorpedodog — COMMENTED** [review]
lots of changes coming in hot

File: abstract_specification.md
Comment (johnkerl) [reply abstract_specification.md:348]: I continue to disagree but won't block the PR.

**Review by johnkerl — APPROVED** [review]

**Comment by thetorpedodog** [comment]
@bkmartinjr @mlin actually I would appreciate if you all took another look particularly for the description of URI semantics that I added.

@johnkerl I clicked you by accident but if you want to give it another ✅ be my guest

File: abstract_specification.md
Comment (bkmartinjr) [inline abstract_specification.md:134]: you might add some wording here that points out that "relative" is an attribute of the particular storage system, eg. S3 really doesn't have directories, but has path prefixes that serve the same purpose.  TileDB-Cloud shows another bunch of special cases if I understand it right.

So I would point out that relative is optional.

And I would reword this to something like:    Where relative URIs are supported by the storage system, collections referencing children by a relative URI will always resolve the child with respect to the parent (base) URI.  For example, on a Posix file system, if the entire directory storing a collection .... <same as you have>

File: abstract_specification.md
Comment (bkmartinjr) [reply abstract_specification.md:501]: we should eventually move to an IDL-like definition if we want the spec to be durable.  But as John says, that is future work, not immediately needed (IMHO)

**Review by bkmartinjr — APPROVED** [review]
One suggestion for a clarification around relative URI resolution in collections.  Otherwise LGTM

File: abstract_specification.md
Comment (thetorpedodog) [reply abstract_specification.md:137]: Added some description of this, along with language indicating that we treat “fake” hierarchies just like normal hierarchies.

**Review by johnkerl — APPROVED** [review]

**Comment by thetorpedodog** [comment]
@atolopko-czi @aaronwolen since this is a spec change we do want unanimous approvals on this so please take a look. Thanks!

**Comment by johnkerl** [comment]
@thetorpedodog ✅  ✅

**Review by aaronwolen — APPROVED** [review]
This is really great

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:483]: Also worth mentioning they inherit the parent's context? Or do we consider that implementation-specific?

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:542]: Does the spec have an opinion on what happens to data originally stored by the object that is replaced at a specified key? Is it orphaned or removed?

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:556]:
```suggestion
Each <code>add_new\_<var>object_type</var></code> method creates a new SOMA dataset in storage, adds it to the collection (with the same semantics as the `create` operation), and returns it to the user. The newly-created entry has the same `context` value as the existing collection and is [owned by the current collection](#operation-close-collection-types).
```

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:567]: Do this need to mention that relative URIs cannot be "multi-level"? I think that's what the spec mentions above.

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:573]: Completely out-of-scope for this PR, but I'm now just wondering why we didn't consider having explicit methods for the various types of collections? `add_new_experiment`, `add_new_measurement`, `add_new_collection`.

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:591]: `Read a subset of data iteratively from the SOMADataFrame`?  Whether or not one is reading a subset, the iterative nature of reads seems like a core part of the design, even at the spec level, given the returns type for DataFrame.read (and SparseNDArray.read())

While most people bothering to read a spec like this will understand that a subset includes the whole set, do we want to rely upon reader knowing that? "Read all or a subset..."

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:776]: important to use `slice` here instead of `subset` as we do above? If it's a technical term we explicitly want, ok.

File: abstract_specification.md
Comment (atolopko-czi) [inline abstract_specification.md:951]:
```suggestion
open(string uri, ...) -> SOMAObject      # identify a SOMA object and open it
```
?

**Review by atolopko-czi — APPROVED** [review]
Really great to read the official details on open/close, read/write, etc.! A number of very minor questions and suggestions. Approving as is, since I found no blockers, but have a look at my comments before merging.

**Review by mlin — APPROVED** [review]

File: abstract_specification.md
Comment (thetorpedodog) [reply abstract_specification.md]: Updated to represent `.collection-info` as a real file and present its contents differently from other paths.

**Pull request merged**
